### PR TITLE
Add recursive tail call test

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #2503 

Tests: #2497
Tests: #2496

## Description

Add test to measure maximum tail call depth.
Fix ref-count leak when maximum tail call depth is reached.

## Testing

New tests + CI/CD.

## Documentation

No.

### CODE REVIEW

File: tests/unit/libbpf_test.cpp [hunk 1]
```diff
@@ -2743,3 +2747,114 @@ TEST_CASE("libbpf_load_stress", "[libbpf]")
...
+
+    // Load the BPF program.
+    REQUIRE(bpf_object__load(object) == 0);
+
+    struct bpf_program* program = bpf_object__find_program_by_name(object, "recurse");
+    REQUIRE(program != nullptr);
+
+    // Get the map used to store the next program to call.
+    struct bpf_map* map = bpf_object__find_map_by_name(object, "map");
+
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2765]: maybe add
```
REQUIRE(map != nullptr);
```

File: tests/unit/libbpf_test.cpp [hunk 2]
```diff
@@ -2743,3 +2747,114 @@ TEST_CASE("libbpf_load_stress", "[libbpf]")
...
+    // Get the map used to record the number of times the program was called.
+    struct bpf_map* canary_map = bpf_object__find_map_by_name(object, "canary");
+
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2768]: perhaps add
```
REQUIRE(canary_map != nullptr);
```

File: tests/unit/libbpf_test.cpp [hunk 3]
```diff
@@ -2743,3 +2747,114 @@ TEST_CASE("libbpf_load_stress", "[libbpf]")
...
+    // Get the fd for the program.
+    fd_t program_fd = bpf_program__fd(program);
+
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2771]: perhaps `REQUIRE(program_fd >= 0)` ?  and same for 2773 and 2776?

File: tests/unit/libbpf_test.cpp [hunk 4]
```diff
@@ -2743,3 +2747,114 @@ TEST_CASE("libbpf_load_stress", "[libbpf]")
...
+{
+    _test_helper_libbpf test_helper;
+    struct bpf_object* object = bpf_object__open("tail_call_sequential_um.dll");
+    REQUIRE(object != nullptr);
+
+    // Load the BPF program.
+    REQUIRE(bpf_object__load(object) == 0);
+
+    // Get the map used to store the next program to call.
+    struct bpf_map* map = bpf_object__find_map_by_name(object, "map");
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2813]: same suggestions about additional REQUIRE's here

File: tests/unit/libbpf_test.cpp [hunk 5]
```diff
@@ -17,6 +17,13 @@
 #include <stop_token>
 #include <thread>
 
+#if !defined(MAX_TAIL_CALL_CNT)
+#define MAX_TAIL_CALL_CNT 32
```
Comment (dv-msft) [inline tests/unit/libbpf_test.cpp:21]: Is there a reason why this can't be defined in one of the published 'ebpf-for-windows' headers and include that header here?

Comment (dthaler) [reply tests/unit/libbpf_test.cpp:21]: It's already defined in `include\bpf_helpers.h`

**Review by dthaler — APPROVED** [review]

File: libs/execution_context/ebpf_program.c [hunk 6]
```diff
@@ -1337,7 +1337,8 @@ ebpf_program_set_tail_call(_In_ const ebpf_program_t* next_program)
         return EBPF_INVALID_ARGUMENT;
     }
 
-    if (state->count == MAX_TAIL_CALL_CNT) {
+    if (state->count == (MAX_TAIL_CALL_CNT - 1)) {
```
Comment (shpalani) [inline libs/execution_context/ebpf_program.c:1340]: Any reason for -1? (31)

1. As per doc, Linux support 32 tails calls.
https://www.man7.org/linux/man-pages/man7/bpf-helpers.7.html

2. As per Cilium documentation, support up to 33 tail calls.
  "This limit comes from the 256 bytes per stack frame multiplied by the tail call count limit (33)"

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:1340]: The count represents how many tail calls have already been made. I.e. if this is the 31st tail call (i.e. it's already 32 deep) then fail this one.

File: libs/platform/user/ebpf_platform_user.cpp [hunk 7]
```diff
@@ -1216,6 +1231,17 @@ ebpf_platform_printk(_In_z_ const char* format, va_list arg_list)
...
+
+    std::string output;
+    output.resize(bytes_written);
+
+    vsprintf_s(output.data(), output.size(), format, arg_list);
+    // Remove the trailing null.
+    output.pop_back();
+
+    std::unique_lock<std::mutex> lock(_ebpf_platform_printk_output_lock);
+    _ebpf_platform_printk_output.emplace_back(std::move(output));
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/platform/user/ebpf_platform_user.cpp:1243]: This is a test see if #2497 reproduces.

File: libs/execution_context/ebpf_program.c [hunk 8]
```diff
@@ -1337,7 +1337,8 @@ ebpf_program_set_tail_call(_In_ const ebpf_program_t* next_program)
...
+        EBPF_OBJECT_RELEASE_REFERENCE(&((ebpf_program_t*)next_program)->object);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:1341]: This is a test to see if #2496 reproduces.

File: tests/unit/libbpf_test.cpp (hunk 5 above)
Comment (dv-msft) [reply tests/unit/libbpf_test.cpp:24]: From an ebpf app writer's perspective though, this is still confusing.  I'd expect the inclusion of the 'one true header' to fix such internal compatibility issues.  All the more so for such key definitions.

File: libs/execution_context/ebpf_program.c (hunk 6 above)
Comment (shpalani) [reply libs/execution_context/ebpf_program.c:1340]: Here is my understanding,
1. 1- 31  is still 31 calls.
2. If let's say, the state starts at 0,  0 - 31, it is  32 calls. Which means 31st call should be allowed.

**Review by dv-msft — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]
